hooks/tk-multi-publish2/basic/collector.py

```diff
@@ -153,180 +192,128 @@ def collect_current_houdini_session(self, settings, parent_item):
         self.logger.info("Collected current Houdini session")
         return session_item
 
-    def collect_node_outputs(self, parent_item):
+    def collect_node_outputs(self, settings, parent_item):
         """
         Creates items for known output nodes
 
         :param parent_item: Parent Item instance
         """
 
+        engine = self.parent.engine
+
         for node_category in _HOUDINI_OUTPUTS:
             for node_type in _HOUDINI_OUTPUTS[node_category]:
+                # get all the nodes for the category and type
+                nodes = hou.nodeType(node_category, node_type).instances()
```

Comment:
You could try refactor the `instances()` fetching out into a separate generator function at the top of the file after the `_HOUDINI_OUTPUTS` statement, e.g.

```python
def _iter_output_nodes():
    """Iterate over all output nodes in the scene.

    For a more functional programming approach, you can try::

        for category, type_names in _HOUDINI_OUTPUTS.items():
            node_types = (hou.nodeType(category, name) for name in type_names)
            for node_type in filter(None, node_types):  # Strip None node_type
                yield category, node_type, node_type.instances() or ()

    :return: Node category, node type and all node instances.
    :rtype: Generator[hou.NodeTypeCategory, hou.NodeType, tuple[hou.Node]]
    """
    for category, type_names in _HOUDINI_OUTPUTS.items():
        for name in type_names:
            node_type = hou.nodeType(category, name)
            if node_type is not None:
                yield category, node_type, node_type.instances() or ()
```

Then you can replace most of this, up to and including line 238, with something like:

```python
        for _, node_type, nodes in _iter_output_nodes():
            type_name = node_type.name()
            get_output_paths_and_templates = None
            # iterate over each node
            for node in nodes:
                if get_output_paths_and_templates is None:
                    get_output_paths_and_templates = getattr(
                        engine.node_handler(node),
                        'get_output_paths_and_templates',
                        None,
                    )
                    if get_output_paths_and_templates is None:
                        # This isn't an export node or missing handler
                        self.logger.info("%s node: %s", type_name, node.path())
                        break

                # get the evaluated path parm value
                self.logger.info(
                    "Processing %s node: %s", type_name, node.path()
                )
                paths_and_templates = get_output_paths_and_templates(node)
                node_item = parent_item.create_item(
                    "houdini.node.{}".format(type_name),
                    "",
                    node.name()
                )
```